src/oph/ehoks/palaute/opiskelija.clj

```diff
@@ -40,95 +41,98 @@
-              (added? :sahkoposti prev-hoks hoks)
-              (log/info msg "`sahkoposti` has been added.")
-
-              (added? :puhelinnumero prev-hoks hoks)
-              (log/info msg "`puhelinnumero` has been added.")
-
-              :else true))))))) ; will be converted to `false`.
+  [field current-hoks updated-hoks]
+  (and (some? current-hoks)
+       (not (get current-hoks field))
+       (some? (get updated-hoks field))))
+
+(defn initial-palaute-state-and-reason
+  "Runs several checks against HOKS and opiskeluoikeus to determine if
+  opiskelijapalautekysely should be initiated.  Returns the initial state
+  of the palaute (or nil if it cannot be formed at all), the field the
+  decision was based on, and the reason for picking that state."
+  ([kysely prev-hoks hoks opiskeluoikeus]
+    (initial-palaute-state-and-reason kysely prev-hoks hoks opiskeluoikeus nil))
+  ([kysely prev-hoks hoks opiskeluoikeus existing-heratteet]
+    (let [herate-basis (herate-date-basis kysely)
+          herate-date (get hoks herate-basis)]
+      (cond
+        (not herate-date)
+        [nil herate-basis :ei-ole]
+
+        (not (palaute/valid-herate-date? herate-date))
+        [:ei-laheteta herate-basis :eri-rahoituskaudella]
+
+        (not (:osaamisen-hankkimisen-tarve hoks))
+        [:ei-laheteta :osaamisen-hankkimisen-tarve :ei-ole]
+
+        (not-any? suoritus/ammatillinen? (:suoritukset opiskeluoikeus))
+        [:ei-laheteta :opiskeluoikeus-oid :ei-ammatillinen]
+
+        (opiskeluoikeus/in-terminal-state? opiskeluoikeus herate-date)
+        [:ei-laheteta :opiskeluoikeus-oid :opiskelu-paattynyt]
+
+        (palaute/feedback-collecting-prevented? opiskeluoikeus herate-date)
+        [:ei-laheteta :opiskeluoikeus-oid :ulkoisesti-rahoitettu]
+
+        (c/tuva-related-hoks? hoks)
+        [:ei-laheteta :tuva-opiskeluoikeus-oid :tuva-opiskeluoikeus]
+
+        (opiskeluoikeus/is-tuva? opiskeluoikeus)
+        [:ei-laheteta :opiskeluoikeus-oid :tuva-opiskeluoikeus]
+
+        (opiskeluoikeus/linked-to-another? opiskeluoikeus)
+        [:ei-laheteta :opiskeluoikeus-oid :liittyva-opiskeluoikeus]
+
+        (palaute/already-initiated? existing-heratteet)
+        [nil herate-basis :jo-lahetetty-talla-rahoituskaudella]
+
+        (and (= kysely :aloituskysely)
+             (added? :osaamisen-hankkimisen-tarve prev-hoks hoks))
+        [:odottaa-kasittelya :osaamisen-hankkimisen-tarve :lisatty]
+
+        (and (= kysely :aloituskysely) (added? :sahkoposti prev-hoks hoks))
+        [:odottaa-kasittelya :sahkoposti :lisatty]
+
+        (and (= kysely :aloituskysely) (added? :puhelinnumero prev-hoks hoks))
+        [:odottaa-kasittelya :puhelinnumero :lisatty]
+
+        (and (= kysely :paattokysely)
+             (added? :osaamisen-saavuttamisen-pvm prev-hoks hoks))
+        [:odottaa-kasittelya :osaamisen-saavuttamisen-pvm :lisatty]
+
+        (and (some? prev-hoks) (= kysely :aloituskysely))
+        [:ei-laheteta nil :ei-muutosta]
```

Comment:
Aloin nyt miettimään, että tarvitaankos näitä tarkistuksia ollenkaan, jos olemassa olevan herätteen tarkistus `palaute/already-initiated?` tehdään yllä ennen näitä? HOKSin ja kyselyn taustatietoja voi muuttaa siihen saakka kunnes palautteen tila on jotain muuta kuin `odottaa_kasittelya` tai `ei_laheteta`. Näiden tarkistusten kannalta taitaa siis olla aika yhdentekevää, onko palaute ehditty jo käsitellä vai ei, koska jos on ehditty käsitellä, `palaute/already-initiated?` yllä estää uudelleenmuodostuksen, ja jos ei ole ehditty käsitellä, palaute menee `odottaa_kasittelya` tilaan ilman näitä tarkistuksia, johon se menisi muutenkin näiden tarkistusten kanssa (miinus tuo `:ei-muutosta`, mikä lienee vähän turha).

Reply:
Totta, hyvä huomio! Tarkistukset siis yksinkertaistuvat mukavasti.

Mielestäni Paula sanoi että sekä aloitus- että päättökysely pitäisi muodostua vaikka opiskelijalla ei olisi mitään yhteystietoja. Tällöin hän voi käydä vastaamassa kyselyyn oman opintopolun -> eHoksin oppijan näkymän kautta. Tän voisi vielä varmistaa kohta vaikka weeklyssä.